### PR TITLE
Production order delivery

### PR DESCRIPTION
**Pull request by amitaibu**
#86

### CODE REVIEW

File: garmentbox/behat/features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -31,8 +36,8 @@ public function __construct(array $parameters) {
   public function iAmLoggedInAs($username) {
     try {
       $password = $this->drupal_users[$username];
-    } catch (Exception $e) {
-      throw new Exception("Password not found for '$username'.");
+    } catch (\Exception $e) {
```
Comment (amitaibu) [inline garmentbox/behat/features/bootstrap/FeatureContext.php]: Line break

File: garmentbox/behat/features/bootstrap/FeatureContext.php [hunk 2]
```diff
@@ -528,4 +533,74 @@ public function iClickTheRowOf($value_in_row) {
...
+      return;
+    }
+
+    throw new \Exception('Row not found.');
+  }
+
+  /**
+   * @Given /^I am on the "([^"]*)" page$/
+   */
+  public function iAmOnThePage($page_name) {
```
Comment (amitaibu) [inline garmentbox/behat/features/bootstrap/FeatureContext.php]: Remove.

File: garmentbox/behat/features/order_delivery.feature [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Feature: Production order delivery.
+  Test the production order delivery flow.
```
Comment (amitaibu) [inline garmentbox/behat/features/order_delivery.feature]: - Rename to "production_delivery"
- Add a bit more info:
  `
  Allow user to mark all the items that were received from production.
  `

File: garmentbox/modules/custom/garmentbox_factory/garmentbox_factory.module [hunk 4]
```diff
@@ -504,6 +529,60 @@ function garmentbox_factory_production_order_items_data($op, $item_variant_nids,
...
+  return !empty($result['node']) ? array_keys($result['node']) : array();
+}
+
+/**
+ * Build a mini inventory line form.
+ *
+ * @param $nid
+ *   Item variant node ID.
+ */
+function _garmentbox_factory_production_order_items_sizes_inputs($nid) {
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/garmentbox_factory.module]: Remove.

File: garmentbox/modules/custom/garmentbox_factory/garmentbox_factory.module [hunk 5]
```diff
@@ -541,10 +632,17 @@ function garmentbox_factory_create_inventory_line($production_order_nid, $item_v
 
   $wrapper = entity_metadata_wrapper('node', $node);
   $wrapper->field_production_order->set($production_order_nid);
+  $wrapper->field_origin_entity->set($origin_nid);
   $wrapper->field_item_variant->set($item_variant_nid);
+  if ($defective) {
+    $wrapper->field_defective->set($defective);
+  }
+  if ($missing) {
+    $wrapper->field_missing_on_il->set($missing);
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/garmentbox_factory.module]: `field_missing_inventory_line`

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+ * Edit form.
+ */
+function garmentbox_factory_production_order_delivery_content_type_edit_form($form, &$form_state) {
+  return $form;
+}
+
+/**
+ * Theme the production order delivery form.
+ */
+function theme_garmentbox_factory_production_order_delivery_form($variables) {
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc:40]: Replace module name with garmentbox_production

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+      }
+
+      $row = _garmentbox_factory_production_order_delivery_form_row($nid, 'received', t('Received'), NULL, $variant['received']);
+      // Override the first row's settings.
+      $row['class'] = array('received');
+      if (!empty($variant["received_variant_$nid"]['#disabled'])) {
+        $row['class'][] = 'disabled';
+      }
+
+      unset($row['ref']);
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: Remove.

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+      );
+      $row['data'][1] = render($variant['title']);
+      $rows[] = $row;
+      $row = _garmentbox_factory_production_order_delivery_form_row($nid, 'original', t('Original'), 'subrow', $variant['original']);
+      $row['data'][1] = render($variant['lines_expander']);
+      $rows[] = $row;
+
+      foreach ($variant['lines'] as $il_nid => $il) {
+        if (!is_numeric($il_nid)) {
+          // Ignore FAPI settings.
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: `
      $keys = array_keys($form);
      element_children($keys);
`

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 9]
```diff
@@ -0,0 +1,672 @@
...
+        }
+
+        $wrapper = entity_metadata_wrapper('node', $il_nid);
+        $label = render($il["il_{$il_nid}_title_variant_{$nid}"]);
+
+        // Add an editable hidden missing row.
+        $rows[] = _garmentbox_factory_production_order_delivery_form_row($nid, 'il', $label, NULL, $il);
+        // Add the "Create missing IL" checkbox.
+        $row = _garmentbox_factory_production_order_delivery_form_row($nid, 'missing', t('Missing'), NULL);
+        $row['data-il-ref'] = $il_nid;
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: `data-il-ref` => `data-inventory-line`

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 10]
```diff
@@ -0,0 +1,672 @@
...
+            continue;
+          }
+
+          $row = _garmentbox_factory_production_order_delivery_form_row($nid, 'missing_created', t('Missing'), NULL, $il);
+          $row['class'] = array('missing-created');
+          $rows[] = $row;
+        }
+      }
+
+      $rows[] = _garmentbox_factory_production_order_delivery_form_row($nid, 'defective', t('Defective'), 'subrow', $variant['defective']);
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: Move t() logic inside _garmentbox_factory_production_order_delivery_form_row()

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+
+  return $row;
+}
+
+/**
+ * Create the production order delivery update form.
+ *
+ * @param $node
+ *   The production order node.
+ */
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc:207]: Add `@see theme_...`

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+      '#disabled' => $variant_data['in_warehouse'],
+    );
+
+    $form['variants'][$nid]['title'] = array(
+      '#markup' => check_plain($variant_data['title']),
+    );
+
+    // Add the lines expander link.
+    $form['variants'][$nid]['lines_expander'] = array(
+      '#markup' => l('', '', array('attributes' => array('class' => array('expander', 'collapsed')))),
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: `<a href="#" class="...">`

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+      }
+    }
+  }
+
+  $totals = array(
+    'total_items_new' => t('Total items new'),
+    'production_cost_new' => t('Production cost new'),
+  );
+  foreach ($totals as $name => $title) {
+    $form['totals'][$name] = array(
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc:338]: Add `#description` with the original values.

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+
+  $form['#attached']['js'] = array(
+    drupal_get_path('module', 'garmentbox_factory') . '/plugins/content_types/production_order_delivery/production_order_delivery.js',
+    array('data' => array(
+      'garmentbox_factory' => array('delivery_data' => $data)),
+      'type' => 'setting',
+    ),
+    drupal_get_path('module', 'format_number') . '/format_number.js',
+    array(
+      'data' => array('format_number' => format_number_get_options()),
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc:360]: Verify we get only a single option, correct for the user.

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+      }
+    }
+
+    $il_nids = array_keys($variant_data['original']['lines']);
+    foreach (node_load_multiple($il_nids) as $node) {
+      // The revision must be enabled before the quantity items are re-saved.
+      $node->revision = TRUE;
+      $original_revision = $node->vid;
+      $missing_size_data = array();
+      // Flag specifying whether the currrent IL has missing quantity.
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: IL => inventory line

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 16]
```diff
@@ -0,0 +1,672 @@
...
+      $il_missing_data = FALSE;
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: => il_has_missing

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 17]
```diff
@@ -0,0 +1,672 @@
...
+
+      $wrapper = entity_metadata_wrapper('node', $node);
+
+      // Reduce quantities of iLines where the received quantity was lower than
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: iLines

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 18]
```diff
@@ -0,0 +1,672 @@
...
+        if ($missing_quantity) {
+          $il_missing_data = TRUE;
+        }
+      }
+
+      // Set the "Current production" inventory lines type to "Regular stock".
+      if ($wrapper->field_inventory_type->getIdentifier() == $current_prooduction_tid) {
+        $wrapper->field_inventory_type->set($regular_stock_tid);
+        // Save the IL revision when the when it was received.
+        $wrapper->field_original_revision->set($original_revision);
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc:495]: @bricel, Please add Behat test...

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 19]
```diff
@@ -0,0 +1,672 @@
...
+          $order_wrapper->save();
+        }
+      }
+    }
+
+    // Check if new inventory lines should be created for the received items.
+    $quantity_info = array();
+    foreach ($variant_data['original']['sizes'] as $tid => $original) {
+      $key = "received_quantity_variant_{$variant_nid}_tid_{$tid}";
+      if (!array_key_exists($key, $form_state['values'])) {
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: array_key_exists() => isset()

File: garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc [hunk 20]
```diff
@@ -0,0 +1,672 @@
...
+  $related_il_nids = garmentbox_factory_get_production_originated_inventory_lines($node->nid);
+  foreach (node_load_multiple($related_il_nids) as $il_node) {
+    $wrapper = entity_metadata_wrapper('node', $il_node);
+    $original_vid = $wrapper->field_original_revision->value();
+    if ($original_vid) {
+      $il_node = node_load(NULL, $original_vid);
+      $wrapper = entity_metadata_wrapper('node', $il_node);
+    }
+
+    // TODO: Are there more possibilities?
```
Comment (amitaibu) [inline garmentbox/modules/custom/garmentbox_factory/plugins/content_types/production_order_delivery/production_order_delivery.inc]: Add exception if type is empty.

**Comment by bitamar** [comment]
Creating a new pull request.

**Pull request closed without merging**